File: src/ollie/context.py
"""Active workflow / interaction context for auto-instrumentation attach."""

from __future__ import annotations

from contextvars import ContextVar, Token
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ollie.interaction_v2 import InteractionHandle
    from ollie.workflow import WorkflowSession

_active_workflow: ContextVar[WorkflowSession | None] = ContextVar("ollie_active_workflow", default=None)
_active_interaction: ContextVar[InteractionHandle | None] = ContextVar(
    "ollie_active_interaction", default=None
)


def get_active_workflow() -> WorkflowSession | None:
    return _active_workflow.get()


def get_active_interaction() -> InteractionHandle | None:
    return _active_interaction.get()


def get_active_parent() -> InteractionHandle | None:
    """Parent for auto-recorded interactions: current open interaction, else workflow root."""
    current = _active_interaction.get()
    if current is not None and not current._closed:
        return current
    wf = _active_workflow.get()
    if wf is not None and wf._root is not None and not wf._root._closed:
        return wf._root
    return None


def set_active_workflow(workflow: WorkflowSession | None) -> Token:
    return _active_workflow.set(workflow)


def reset_active_workflow(token: Token) -> None:
    _active_workflow.reset(token)


def set_active_interaction(handle: InteractionHandle | None) -> Token:
    return _active_interaction.set(handle)


def reset_active_interaction(token: Token) -> None:
    _active_interaction.reset(token)

File: packages/ts/src/context.ts
import { AsyncLocalStorage } from "node:async_hooks";
import type { InteractionHandle } from "./interaction.js";
import type { WorkflowSession } from "./workflow.js";

interface OllieContext {
  workflow: WorkflowSession | null;
  interaction: InteractionHandle | null;
}

const storage = new AsyncLocalStorage<OllieContext>();

function getStore(): OllieContext {
  return storage.getStore() ?? { workflow: null, interaction: null };
}

export function getActiveWorkflow(): WorkflowSession | null {
  return getStore().workflow;
}

export function getActiveInteraction(): InteractionHandle | null {
  return getStore().interaction;
}

export function getActiveParent(): InteractionHandle | null {
  const current = getActiveInteraction();
  if (current && !current.closed) return current;
  const wf = getActiveWorkflow();
  if (wf?.root && !wf.root.closed) return wf.root;
  return null;
}

export function runWithWorkflow<T>(workflow: WorkflowSession | null, fn: () => T): T {
  const prev = getStore();
  return storage.run({ ...prev, workflow }, fn);
}

export function runWithInteraction<T>(interaction: InteractionHandle | null, fn: () => T): T {
  const prev = getStore();
  return storage.run({ ...prev, interaction }, fn);
}

export function setActiveWorkflow(workflow: WorkflowSession | null): void {
  const prev = getStore();
  storage.enterWith({ ...prev, workflow });
}

export function setActiveInteraction(interaction: InteractionHandle | null): void {
  const prev = getStore();
  storage.enterWith({ ...prev, interaction });
}
